File: Pinball App 2/Pinball App 2/practice/ScoreScannerModels.swift
import Foundation
import CoreGraphics
import SwiftUI

enum ScoreScannerDisplayMode: String, CaseIterable, Identifiable {
    case lcd
    case dmd
    case segmented

    var id: String { rawValue }
}

enum ScoreScannerStatus: Equatable {
    case cameraPermissionRequired
    case cameraUnavailable
    case searching
    case reading
    case stableCandidate
    case locked
    case failedNoReading

    var title: String {
        switch self {
        case .cameraPermissionRequired:
            return "Camera access required"
        case .cameraUnavailable:
            return "Camera unavailable"
        case .searching:
            return "Searching"
        case .reading:
            return "Reading"
        case .stableCandidate:
            return "Stable candidate"
        case .locked:
            return "Locked"
        case .failedNoReading:
            return "No reading"
        }
    }

    var detail: String {
        switch self {
        case .cameraPermissionRequired:
            return "Allow camera access to scan score displays on-device."
        case .cameraUnavailable:
            return "This device could not start the rear camera."
        case .searching:
            return "Align the score display inside the box."
        case .reading:
            return "Live OCR is tracking the display."
        case .stableCandidate:
            return "Hold steady for a clean lock."
        case .locked:
            return "Stable reading captured. Confirm or edit before use."
        case .failedNoReading:
            return "No stable numeric reading yet. Freeze and confirm manually if needed."
        }
    }
}

struct ScoreOCRObservation: Equatable {
    let text: String
    let confidence: Float
    let boundingBox: CGRect
}

struct ScoreScannerCandidate: Equatable {
    let rawText: String
    let normalizedScore: Int
    let formattedScore: String
    let confidence: Float
    let boundingBox: CGRect
    let digitCount: Int
    let centerBias: Double
}

struct ScoreOCRAnalysis: Equatable {
    let bestCandidate: ScoreScannerCandidate?
    let candidates: [ScoreScannerCandidate]
}

struct ScoreScannerLockedReading: Equatable {
    let score: Int
    let formattedScore: String
    let rawText: String
    let confidence: Float
    let averageConfidence: Float
}

struct ScoreScannerPreviewMapping: Equatable {
    let previewBounds: CGRect
    let targetRect: CGRect
}

enum ScoreScannerFrameMapper {
    static func cropRect(frameExtent: CGRect, previewMapping: ScoreScannerPreviewMapping) -> CGRect? {
        guard frameExtent.width > 0, frameExtent.height > 0 else { return nil }

        let previewBounds = previewMapping.previewBounds.standardized
        let targetRect = previewMapping.targetRect.intersection(previewBounds).standardized
        guard previewBounds.width > 0,
              previewBounds.height > 0,
              !targetRect.isNull,
              !targetRect.isEmpty else {
            return nil
        }

        let scale = max(
            previewBounds.width / frameExtent.width,
            previewBounds.height / frameExtent.height
        )
        guard scale > 0 else { return nil }

        let displayedSize = CGSize(
            width: frameExtent.width * scale,
            height: frameExtent.height * scale
        )
        let imageRectInPreview = CGRect(
            x: previewBounds.minX + ((previewBounds.width - displayedSize.width) / 2),
            y: previewBounds.minY + ((previewBounds.height - displayedSize.height) / 2),
            width: displayedSize.width,
            height: displayedSize.height
        )
        let visibleRect = targetRect.intersection(imageRectInPreview)
        guard !visibleRect.isNull, !visibleRect.isEmpty else { return nil }

        let xInImage = (visibleRect.minX - imageRectInPreview.minX) / scale
        let yInImageTop = (visibleRect.minY - imageRectInPreview.minY) / scale
        let widthInImage = visibleRect.width / scale
        let heightInImage = visibleRect.height / scale

        let x = frameExtent.minX + xInImage
        let y = frameExtent.minY + (frameExtent.height - yInImageTop - heightInImage)
        let cropRect = CGRect(x: x, y: y, width: widthInImage, height: heightInImage)
            .integral
            .intersection(frameExtent)

        guard cropRect.width > 0, cropRect.height > 0 else { return nil }
        return cropRect
    }

    static func cropRect(frameExtent: CGRect, normalizedRect: CGRect) -> CGRect? {
        guard frameExtent.width > 0, frameExtent.height > 0 else { return nil }

        let width = frameExtent.width * normalizedRect.width
        let height = frameExtent.height * normalizedRect.height
        let x = frameExtent.minX + (frameExtent.width * normalizedRect.minX)
        let y = frameExtent.minY + (frameExtent.height * (1 - normalizedRect.maxY))
        let cropRect = CGRect(x: x, y: y, width: width, height: height)
            .integral
            .intersection(frameExtent)

        guard cropRect.width > 0, cropRect.height > 0 else { return nil }
        return cropRect
    }
}

struct ScoreScannerTargetBoxLayout {
    static func rect(in size: CGSize, safeAreaInsets: EdgeInsets) -> CGRect {
        let width = min(size.width * 0.82, 420)
        let height = min(max(size.height * 0.10, 78), 104)
        let x = (size.width - width) / 2
        let topInset = safeAreaInsets.top + 192
        let y = min(max(topInset, 190), max(size.height * 0.26, topInset))
        return CGRect(x: x, y: y, width: width, height: height)
    }

    static let fallbackNormalizedRect = CGRect(x: 0.12, y: 0.27, width: 0.76, height: 0.13)
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/practice/ScoreScannerModels.kt
package com.pillyliu.pinprofandroid.practice

import android.graphics.RectF
import androidx.compose.ui.geometry.Size
import kotlin.math.max
import kotlin.math.min

internal enum class ScoreScannerDisplayMode {
    Lcd,
    Dmd,
    Segmented,
}

internal enum class ScoreScannerStatus(
    val title: String,
    val detail: String,
) {
    CameraPermissionRequired(
        title = "Camera access required",
        detail = "Allow camera access to scan score displays on-device.",
    ),
    CameraUnavailable(
        title = "Camera unavailable",
        detail = "This device could not start the rear camera.",
    ),
    Searching(
        title = "Searching",
        detail = "Align the score display inside the box.",
    ),
    Reading(
        title = "Reading",
        detail = "Live OCR is tracking the display.",
    ),
    StableCandidate(
        title = "Stable candidate",
        detail = "Hold steady for a clean lock.",
    ),
    Locked(
        title = "Locked",
        detail = "Stable reading captured. Confirm or edit before use.",
    ),
    FailedNoReading(
        title = "No reading",
        detail = "No stable numeric reading yet. Freeze and confirm manually if needed.",
    ),
}

internal data class ScoreOcrObservation(
    val text: String,
    val confidence: Float,
    val boundingBox: RectF,
)

internal data class ScoreScannerCandidate(
    val rawText: String,
    val normalizedScore: Int,
    val formattedScore: String,
    val confidence: Float,
    val boundingBox: RectF,
    val digitCount: Int,
    val centerBias: Double,
)

internal data class ScoreScannerAnalysis(
    val bestCandidate: ScoreScannerCandidate?,
    val candidates: List<ScoreScannerCandidate>,
)

internal data class ScoreScannerLockedReading(
    val score: Int,
    val formattedScore: String,
    val rawText: String,
    val confidence: Float,
    val averageConfidence: Float,
)

internal data class ScoreScannerPreviewMapping(
    val previewBounds: RectF,
    val targetRect: RectF,
)

internal object ScoreScannerFrameMapper {
    fun cropRect(
        frameSize: Size,
        previewMapping: ScoreScannerPreviewMapping,
    ): RectF? {
        if (frameSize.width <= 0f || frameSize.height <= 0f) return null

        val previewBounds = previewMapping.previewBounds.standardized()
        val targetRect = previewMapping.targetRect.standardized().intersecting(previewBounds) ?: return null
        if (previewBounds.width() <= 0f || previewBounds.height() <= 0f) return null

        val scale = max(
            previewBounds.width() / frameSize.width,
            previewBounds.height() / frameSize.height,
        )
        if (scale <= 0f) return null

        val displayedWidth = frameSize.width * scale
        val displayedHeight = frameSize.height * scale
        val imageRectInPreview = RectF(
            previewBounds.left + ((previewBounds.width() - displayedWidth) / 2f),
            previewBounds.top + ((previewBounds.height() - displayedHeight) / 2f),
            previewBounds.left + ((previewBounds.width() - displayedWidth) / 2f) + displayedWidth,
            previewBounds.top + ((previewBounds.height() - displayedHeight) / 2f) + displayedHeight,
        )
        val visibleRect = targetRect.intersecting(imageRectInPreview) ?: return null

        val xInImage = (visibleRect.left - imageRectInPreview.left) / scale
        val yInImageTop = (visibleRect.top - imageRectInPreview.top) / scale
        val widthInImage = visibleRect.width() / scale
        val heightInImage = visibleRect.height() / scale

        val x = xInImage
        val y = frameSize.height - yInImageTop - heightInImage
        return RectF(
            x,
            y,
            x + widthInImage,
            y + heightInImage,
        ).standardized().intersectWithin(frameSize.width, frameSize.height)
    }

    fun cropRect(
        frameSize: Size,
        normalizedRect: RectF,
    ): RectF? {
        if (frameSize.width <= 0f || frameSize.height <= 0f) return null

        val width = frameSize.width * normalizedRect.width()
        val height = frameSize.height * normalizedRect.height()
        val x = frameSize.width * normalizedRect.left
        val y = frameSize.height * (1f - normalizedRect.bottom)
        return RectF(x, y, x + width, y + height).intersectWithin(frameSize.width, frameSize.height)
    }

    val fallbackNormalizedRect = RectF(0.12f, 0.27f, 0.88f, 0.40f)
}

internal fun RectF.standardized(): RectF {
    val left = min(this.left, this.right)
    val top = min(this.top, this.bottom)
    val right = max(this.left, this.right)
    val bottom = max(this.top, this.bottom)
    return RectF(left, top, right, bottom)
}

internal fun RectF.intersecting(other: RectF): RectF? {
    val left = max(this.left, other.left)
    val top = max(this.top, other.top)
    val right = min(this.right, other.right)
    val bottom = min(this.bottom, other.bottom)
    if (right <= left || bottom <= top) return null
    return RectF(left, top, right, bottom)
}

internal fun RectF.intersectWithin(width: Float, height: Float): RectF? {
    val bounded = intersecting(RectF(0f, 0f, width, height)) ?: return null
    if (bounded.width() <= 0f || bounded.height() <= 0f) return null
    return bounded
}
